crates/uv-distribution/src/lib.rs

```diff
@@ -15,20 +25,118 @@ mod error;
 mod git;
 mod index;
 mod locks;
+pub mod pyproject;
 mod reporter;
+mod requirement_lowering;
 mod source;
+mod workspace;
+
+#[derive(Debug, Error)]
+pub enum MetadataLoweringError {
+    #[error(transparent)]
+    Workspace(#[from] WorkspaceError),
+    #[error(transparent)]
+    Lowering(#[from] LoweringError),
+}
+
+#[derive(Debug, Clone)]
+pub struct Metadata {
+    // Mandatory fields
+    pub name: PackageName,
+    pub version: Version,
+    // Optional fields
+    pub requires_dist: Vec<pypi_types::Requirement>,
+    pub requires_python: Option<VersionSpecifiers>,
+    pub provides_extras: Vec<ExtraName>,
+}
+
+impl Metadata {
+    /// Lower without considering `tool.uv` in `pyproject.toml`, used for index and other archive
+    /// dependencies.
+    pub fn from_metadata23(metadata: Metadata23) -> Self {
+        Self {
+            name: metadata.name,
+            version: metadata.version,
+            requires_dist: metadata
+                .requires_dist
+                .into_iter()
+                .map(pypi_types::Requirement::from)
+                .collect(),
+            requires_python: metadata.requires_python,
+            provides_extras: metadata.provides_extras,
+        }
+    }
+
+    /// Lower by considering `tool.uv` in `pyproject.toml` if present, used for Git and directory
+    /// dependencies.
+    pub async fn from_workspace(
```

Comment:
Kind of wish none of this were async. Wondering if it shouldn't be given that it's all fs operations.

Reply:
Like i said in the initial PR, those are only async for consistency, we can totally make them sync